### PR TITLE
Adding more optional data for different rolls

### PR DESCRIPTION
**Pull request by IrishWolf**
This allows you to show additional datas for rolls (should be issue #60).
I tested it with the item-macros and that worked, too.

The settings let you differentiate between characters:
image
image

**Attribute (original / with data)**
image image

**Skill (original / with data)**
image image

**Weapon (original / with data (melee))**
image image

**Automatic Range Weapon (with data (without using automatic) / with data (while using automatic))**
image image

**Explosive (original / with data)**
image image

**Armor (original / with data)**
image image

**Ship-Roll (original / with data)** (Narzalus is the name of the ship)
image image

**New with <1 dice**
image

**New without additional data**
image

**P.S.**
The numbered dice are not part of this!

### CODE REVIEW

**Comment by IrishWolf** [comment]
Maybe @coderunner could help with the french translations?

**Comment by coderunner** [comment]
Also, I would vote for making it the default (without settings, always on). Any reason we would not want to show the info in some cases?

**Comment by IrishWolf** [comment]
added that automatic fire is now part of the Total roll (automatic -2 dice and message for number of rolls)

without automatic fire:
image

with automatic fire ( ignored 1):
image

File: module/actor/actor-sheet.js [hunk 1]
```diff
@@ -477,19 +477,38 @@ export class yzecoriolisActorSheet extends ActorSheet {
...
+        : 0,
       modifier: 0,
-      bonus: dataset.bonus ? Number(dataset.bonus) : 0,
+      bonus: dataset.bonus
+        ? Number(dataset.bonus)
+        : 0,
       rollTitle: dataset.label,
       pushed: false,
-      actorType: this.actor.data.type,
+      weapon_isAutomatic: item?.automatic,
```
Comment (coderunner) [inline module/actor/actor-sheet.js:503]: Detail : For consistency with other field in camel case: `weapon_isAutomatic` => `weaponIsAutomatic` ? 
and similarly for the other attributes.

File: module/settings.js [hunk 2]
```diff
@@ -52,4 +52,106 @@ export const registerSystemSettings = function () {
...
+      "pc": game.i18n.localize("YZECORIOLIS.SettingRollInfoPC"),
+      "npc": game.i18n.localize("YZECORIOLIS.SettingRollInfoNPC"),
+      "all": game.i18n.localize("YZECORIOLIS.SettingRollInfoAll")
+      },
+    default: "no",
+    onChange: applyRollInfoSetting,
+  });
 };
+
+export function applyRollInfoSetting() {
```
Comment (coderunner) [inline module/settings.js:137]: Shouldn't we register these helper in all cases?

Comment (coderunner) [reply module/settings.js:137]: Actually, I think they are registered on init. I don't think we need to register them again `onChange`.

File: module/coriolis-roll.js [hunk 3]
```diff
@@ -413,6 +456,150 @@ function getTooltipData(results) {
...
+  }
+  return AttrKey;
+}
+
+function getRollSkill(rollData) {
+  return `${rollData.skill}`;
+}
+
+function getRollSkillName(rollData) {
+  let SkillKey = "";
```
Comment (coderunner) [inline module/coriolis-roll.js:491]: Couldn't we reuse `YZECORIOLIS.skills` .... like `YZECORIOLIS.skills[rollData.skillKey]`?

File: module/coriolis-roll.js [hunk 4]
```diff
@@ -413,6 +456,150 @@ function getTooltipData(results) {
...
+  return `${rollData.weapon_damage}`;
+}
+
+function getRollDmgText(rollData) {
+  return `${rollData.weapon_damageText}`;
+}
+
+function getRollRange(rollData) {
+  let range = "";
+  switch (rollData.weapon_range) {
```
Comment (coderunner) [inline module/coriolis-roll.js:590]: Same comment as above.

File: lang/en.json [hunk 5]
```diff
@@ -396,5 +410,8 @@
   "YZECORIOLIS.ItemDeleteConfirmationTitle": "Delete Item",
   "YZECORIOLIS.ItemDeleteConfirmationContent": "Are you sure you want to delete this item?",
   "YZECORIOLIS.AutomaticFire": "Automatic fire",
-  "YZECORIOLIS.AutomaticFireIgnoredOnes": "Number of ignored 1s (max 2)"
-}
\ No newline at end of file
+  "YZECORIOLIS.AutomaticFireIgnoredOnes": "Number of ignored 1s (max 2)",
+
+  "YZECORIOLIS.RollTotal": "Total rolled",
+  "YZECORIOLIS.RollSubstitute": "2d6-saving throw instead of"
```
Comment (coderunner) [inline lang/en.json:416]: I think the wording elsewhere in the code is a `desperation roll` so I'd suggest something like `desperation roll (2d6)` maybe?

Comment (IrishWolf) [reply lang/en.json:416]: yes, I will change that :)

File: module/settings.js (hunk 2 above)
Comment (IrishWolf) [reply module/settings.js:137]: True, works without. I will cut it here.

File: module/coriolis-roll.js [hunk 6]
```diff
@@ -324,6 +328,26 @@ async function showChatMessage(chatMsgOptions, resultData) {
...
+    actorType: getActorType(resultData.rollData),
+    rollType: getRollType(resultData.rollData),
+    attribute: getRollAttribute(resultData.rollData),
+    attributeName: getRollAttributeName(resultData.rollData),
+    skill: getRollSkill(resultData.rollData),
+    skillName: getRollSkillName(resultData.rollData),
+    modifier: getRollModifier(resultData.rollData),
+    isAutomatic: getRollIsAuto(resultData.rollData),
+    isAutomaticActive: getRollIsAutoActive(resultData.rollData),
+    AutomaticRollAmount: parseInt(getRollAutoIgnoOnes(resultData.rollData)) + 1,
```
[the remark this answers is not in the document]
Comment (IrishWolf) [reply module/coriolis-roll.js:341]: yes, must have overseen this

File: module/coriolis-roll.js [hunk 7]
```diff
@@ -413,6 +456,150 @@ function getTooltipData(results) {
...
+function getRollBonus(rollData) {
+  return `${rollData.bonus}`;
+}
+
+function getRollBlastPower(rollData) {
+  return `${rollData.weapon_blastPower}`;
+}
+
+function getRollBlastRadius(rollData) {
+  let radius = "";
```
[the remark this answers is not in the document]
Comment (IrishWolf) [reply module/coriolis-roll.js:558]: Yes, I wasn't sure how that is reusable. But I tryed it now till it works. (easy if you understand it >__<)

File: module/coriolis-roll.js (hunk 4 above)
Comment (IrishWolf) [reply module/coriolis-roll.js:590]: Yes, I wasn't sure how that is reusable. But I tryed it now till it works. (easy if you understand it >__<)

File: module/coriolis-roll.js [hunk 8]
```diff
@@ -413,6 +456,150 @@ function getTooltipData(results) {
...
+function getRollType(rollData) {
+  return `${rollData.rollType}`;
+}
+
+function getRollAttribute(rollData) {
+  return `${rollData.attribute}`;
+}
+
+function getRollAttributeName(rollData) {
+  let AttrKey = "";
```
[the remark this answers is not in the document]
Comment (IrishWolf) [reply module/coriolis-roll.js:472]: Yes, I wasn't sure how that is reusable. But I tryed it now till it works. (easy if you understand it >__<)

File: module/coriolis-roll.js (hunk 3 above)
Comment (IrishWolf) [reply module/coriolis-roll.js:491]: Yes, I wasn't sure how that is reusable. But I tryed it now till it works. (easy if you understand it >__<)

**Review by IrishWolf — COMMENTED** [review]
@coderunner I edited some things :D

- settings reduced from five to one
- optimised code (reusing configs etc)
- adjusted wording
- clarified ship-rolls (look at the edit of the entry-post)

Can you check my newly created chaos again? ^^'

**Comment by coderunner** [comment]
@IrishWolf Amazing work! :clap: 👏 :clap: Can't wait to use this!!!  

I'll provide the french translation shortly.

**Comment by coderunner** [comment]
```
"YZECORIOLIS.SettingAdditionalRollInfos": "Afficher les informations supplémentaires lors des jets de dés",
"YZECORIOLIS.SettingAdditionalRollInfosHint": "Afficher les informations supplémentaires (par exemple: points d'attribut, points de compétence, modificateur, bonus, dégâts, critique, portée, automatique, puissance, rayon).",
"YZECORIOLIS.SettingAdditionalRollInfosNo": "Aucune information supplémentaire",
"YZECORIOLIS.SettingAdditionalRollInfosPC": "Uniquement pour les personnages joueurs",
"YZECORIOLIS.SettingAdditionalRollInfosNPC": "Uniquement pour les personnages non-joueurs",
"YZECORIOLIS.SettingAdditionalRollInfosAll": "Pour tous les personnages",
"YZECORIOLIS.RollTotal": "Description du jet",
"YZECORIOLIS.RollDesperation": "Jet de désespoir (2d6)"
 ```

**Comment by IrishWolf** [comment]
@coderunner I just noticed, that the following Settings are not translated in french yet.
If you can provide that too, I would push it with this together.

```
  "YZECORIOLIS.SettingAlwaysShowFeatures": "Always show features",
  "YZECORIOLIS.SettingAlwaysShowFeaturesHint": "Display item features inline for weapons, explosives, and armor in character sheets.",
```

**Comment by coderunner** [comment]
@IrishWolf 
```
"YZECORIOLIS.SettingAlwaysShowFeatures": "Toujours afficher les traits",
"YZECORIOLIS.SettingAlwaysShowFeaturesHint": "Afficher, sur la feuille de personnage, les traits pour les armes, les explosifs et les armures.",
```

**Review by coderunner — APPROVED** [review]

**Comment by AlexOkafor** [comment]
I love these changes btw! 

> I always roll private as a GM so I personally wouldn't bother about half completed npc sheets and on-the-fly modifiers.
> 
> The way I see it
> 
> * this will save GMs and players lots of time checking damage, crit threshold, etc on so many rolls that I'd assume pretty much everyone would want that (like your d6.d6 initiative)
> * more settings = more complexity = more potential bugs... and more translations :)
> 
> @AlexOkafor what do you think?

Totally in line with fewer options the better.  I'd rather have it as the default as it is simply an overall improvement to the chat messages rather than a preferential/aesthetic choice.

**Comment by AlexOkafor** [comment]
@IrishWolf Love what you did here! This is AMAZING! I think the displayed data here is great for PC/NPCs. I'll go ahead and remove the option and display by default if that's ok? 

If I'm missing a use-case that wasn't covered in the above discussion about why you wouldn't want to display all the data, let me know and I can re-introduce the options in a later update.

 I think a future alternative to options is a pass on the UI for the roll message could be done that would allow for rollouts/hovers for those that are wanting to dive more deeply into the roll data.

**Comment by IrishWolf** [comment]
> @IrishWolf Love what you did here! This is AMAZING! I think the displayed data here is great for PC/NPCs. I'll go ahead and remove the option and display by default if that's ok?
> 
> If I'm missing a use-case that wasn't covered in the above discussion about why you wouldn't want to display all the data, let me know and I can re-introduce the options in a later update.
> 
> I think a future alternative to options is a pass on the UI for the roll message could be done that would allow for rollouts/hovers for those that are wanting to dive more deeply into the roll data.

As I wrote I would like the full stats just for the PC and not NPC, cause sometimes I will roll for npc with some halfway unprepared sheets and the players shouldn't see that the numbers are strangely distributed (I don't like rolling private in this regard).
But I guess it's fine this way also as I understand the problems you two mentioned. :)

**Pull request merged**